components/calendar/src/julian.rs

```diff
@@ -0,0 +1,317 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! This module contains types and implementations for the Julian calendar
+
+use crate::iso::{Iso, IsoDateInner, IsoDay, IsoMonth, IsoYear};
+use crate::{types, Calendar, Date, DateDuration, DateDurationUnit, DateTime, DateTimeError};
+use core::convert::TryInto;
+
+#[derive(Copy, Clone, Debug, Default)]
+// The Julian calendar
+pub struct Julian;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+// The inner date type used for representing Date<Georgian>
+pub struct JulianDateInner(IsoDateInner);
+
+impl Calendar for Julian {
+    type DateInner = JulianDateInner;
+    fn date_from_iso(&self, iso: Date<Iso>) -> JulianDateInner {
+        let added_days = Self::calculate_day_difference_between_calendars(*iso.inner());
+        Self::offset_date(
+            &mut JulianDateInner(*iso.inner()),
+            DateDuration::new(0, 0, 0, -added_days),
+        )
+    }
+
+    fn date_to_iso(&self, date: &Self::DateInner) -> Date<Iso> {
+        let added_days = Self::calculate_day_difference_between_calendars(date.0);
```

Comment:
question: I'm unsure how this manages to work _both_ ways, since this function seems to take an ISO date and tell you what to add to it to get the Julian date, the reverse should not automatically work? I guess that's what the edge case is dealing with, but all of this is kinda unclear, can you link to sources or perhaps factor this as two separate functions?

Reply:
I dont really have any documentation on it, wrote the code based on my calculations sort of, so this might as well be wrong.

> perhaps factor this as two separate functions?

Which part specifically?! I think separating out edge case to a separate method might make it harder to read, as the edge case is part of the core logic

Reply:
I guess it's fine if it's in a single function, but please add comments explaining why this is needed

Or ideally look up an algorithm for each direction and implement that.

Reply:
Added comments

Reply:
So I think the thing that concerns me about this is the use of `calculate_day_difference()`: what if it is called around the boundary when the year is different in ISO and Julian and each year has a different day difference.

I would really be more comfortable if there were some sources cited on this.

Reply:
@sffc thoughts?

@pandusonu2 if you need, code from the calendar arithmetic book has been published and the relevant code is https://github.com/EdReingold/calendar-code2/blob/1ee51ecfaae6f856b0d7de3e36e9042100b4f424/calendar.l#L1689-L1738 , though it's in Lisp. It's okay to refer to this code

Reply:
I added test cases for the same, if that helps

Also one of the source I found is https://smallbusiness.chron.com/convert-julian-iso-49137.html perhaps?! but it doesnt handle the edge case.

The above (N3/4 - 2) is equivalent to (n/100 - n/400 - 2) which is being done in the calculate_day_difference()

Reply:
Also, the main resource I used for this conversion is actually: https://www.fourmilab.ch/documents/calendar/ 
I did the conversions to check how the dates operate, and wrote the code based on that

Reply:
Thanks for the above link, https://github.com/EdReingold/calendar-code2/blob/1ee51ecfaae6f856b0d7de3e36e9042100b4f424/calendar.l#L1704

I think that explains the edge case handling I had done in the `date_to_iso()`

Reply:
Can you link to that source and write your math in a way that matches theirs, with explainer comments?

In particular I'm still not convinced that `calculate_day_difference()` calculates the slack for an _iso_ date, can always work when applied to a _julian_ date since it's going to have problems in the period where that slack shifts